src/collision.cpp

```diff
@@ -66,95 +82,94 @@ CollisionAxis axisAlignedCollision(
 	//TimeTaker tt("axisAlignedCollision");
 
 	aabb3f relbox(
-			movingbox.MaxEdge.X - movingbox.MinEdge.X + staticbox.MaxEdge.X - staticbox.MinEdge.X,						// sum of the widths
-			movingbox.MaxEdge.Y - movingbox.MinEdge.Y + staticbox.MaxEdge.Y - staticbox.MinEdge.Y,
-			movingbox.MaxEdge.Z - movingbox.MinEdge.Z + staticbox.MaxEdge.Z - staticbox.MinEdge.Z,
+			(movingbox.MaxEdge.X - movingbox.MinEdge.X) + (staticbox.MaxEdge.X - staticbox.MinEdge.X),						// sum of the widths
+			(movingbox.MaxEdge.Y - movingbox.MinEdge.Y) + (staticbox.MaxEdge.Y - staticbox.MinEdge.Y),
+			(movingbox.MaxEdge.Z - movingbox.MinEdge.Z) + (staticbox.MaxEdge.Z - staticbox.MinEdge.Z),
 			std::max(movingbox.MaxEdge.X, staticbox.MaxEdge.X) - std::min(movingbox.MinEdge.X, staticbox.MinEdge.X),	//outer bounding 'box' dimensions
 			std::max(movingbox.MaxEdge.Y, staticbox.MaxEdge.Y) - std::min(movingbox.MinEdge.Y, staticbox.MinEdge.Y),
 			std::max(movingbox.MaxEdge.Z, staticbox.MaxEdge.Z) - std::min(movingbox.MinEdge.Z, staticbox.MinEdge.Z)
 	);
 
 	const f32 dtime_max = *dtime;
-	const f32 inner_margin = -1.5f;
+	f32 inner_margin;		// the distance of clipping recovery
 	f32 distance;
 	f32 time;
 
-	if (speed.X) {
-		distance = relbox.MaxEdge.X - relbox.MinEdge.X;
 
-		*dtime = distance >= 0 ? std::abs(distance / speed.X) : -std::abs(distance / speed.X);
+	if (speed.Y) {
+		distance = relbox.MaxEdge.Y - relbox.MinEdge.Y;
+		*dtime = distance / std::abs(speed.Y);
 		time = std::max(*dtime, 0.0f);
 
-		if (distance > inner_margin) {
-			if (*dtime <= dtime_max) {
-				if ((speed.X > 0 && staticbox.MaxEdge.X > movingbox.MaxEdge.X) ||
-						(speed.X < 0 && staticbox.MinEdge.X < movingbox.MinEdge.X)) {
-					if (
-						(std::max(movingbox.MaxEdge.Y + speed.Y * time, staticbox.MaxEdge.Y)
-							- std::min(movingbox.MinEdge.Y + speed.Y * time, staticbox.MinEdge.Y)
-							- relbox.MinEdge.Y < 0) &&
+		if (*dtime <= dtime_max) {
+			inner_margin = std::max(-0.5f * (staticbox.MaxEdge.Y - staticbox.MinEdge.Y), -2.0f);
+
+			if ((speed.Y > 0 && staticbox.MinEdge.Y - movingbox.MaxEdge.Y > inner_margin) ||
+				(speed.Y < 0 && movingbox.MinEdge.Y - staticbox.MaxEdge.Y > inner_margin)) {
```

Comment:
Would it be possible to simplify these checks a bit?

```C++
aabb3f moved_box = movingbox;
moved_box.MinEdge *= speed * time;
moved_box.MaxEdge *= speed * time;

f32 direction = (speed.Y == 0.0f) ? 0.0f : speed.Y / abs(speed.Y);

if (staticbox.MinEdge.Y - movingbox.MaxEdge.Y > inner_margin * direction) {
	if (std::max(moved_box.MaxEdge.X, staticbox.MaxEdge.X)
			- std::max(moved_box.MinEdge.X, staticbox.MinEdge.X) < relbox.MinEdge.X)
		// if ( .. Z

```

To avoid code duplication, it would also be possible to make use of a generic function like 
```C++
template <float v3f::*A, float v3f::*B, float v3f::*B>
static bool checkCollision(const v3f &speed, const aabb3f &staticbox, const aabb3f &movingbox)
{
	if (staticbox.MinEdge.*A - movingbox.MaxEdge.*A > inner_margin * direction) {
}
```

Reply:
- simplify: 
I wish but unfortunately the suggested code isn't equivalent.
- generic:
Yes, that probably wouldn't improve readability that much and I'm not terribly familiar with pointers to members so I really prefer not to do that.

Reply:
Which part of the suggested code is not equivalent? Is it the msising `direction != 0.0f` in the first condition? Just wondering what I missed there.